// survey-analytics/fesm/shared.mjs
/*!
 * surveyjs - SurveyJS Dashboard library v2.2.5
 * Copyright (c) 2015-2025 Devsoft Baltic OÜ  - http://surveyjs.io/
 * License: MIT (http://www.opensource.org/licenses/mit-license.php)
 */

// Uncomment the line below if you create a custom dictionary
// import { localization } from "../localizationManager";
var englishStrings = {
    groupButton: "Group By Me",
    ungroupButton: "Ungroup By Me",
    selectButton: "Select Me",
    hideColumn: "Hide column",
    showColumn: "Show column",
    makePrivateColumn: "Make column private",
    makePublicColumn: "Make column public",
    moveToDetail: "Move to Detail",
    showAsColumn: "Show as Column",
    filterPlaceholder: "Search...",
    removeRows: "Remove rows",
    showLabel: "Show",
    entriesLabel: "entries",
    visualizer_text: "Texts in table",
    visualizer_wordcloud: "Wordcloud",
    visualizer_histogram: "Histogram",
    visualizer_number: "Average",
    visualizer_choices: "Table",
    visualizer_selectBase: "Chart",
    visualizer_matrix: "Chart",
    chartType_bar: "Bar",
    chartType_vbar: "Vertical Bar",
    chartType_stackedbar: "Stacked Bar",
    chartType_doughnut: "Doughnut",
    chartType_pie: "Pie",
    chartType_scatter: "Scatter",
    chartType_gauge: "Gauge",
    chartType_bullet: "Bullet",
    chartType_line: "Line",
    hideButton: "Hide",
    makePrivateButton: "Make private",
    makePublicButton: "Make public",
    showButton: "Show",
    filter: "Filter",
    resetFilter: "Reset Filter",
    changeLocale: "Change Locale",
    clearButton: "Clear",
    addElement: "Choose question to show...",
    defaultOrder: "Default Order",
    ascOrder: "Ascending",
    descOrder: "Descending",
    showMinorColumns: "Show minor columns",
    actionsColumn: "Actions",
    otherCommentTitle: "Other items and comments",
    showPercentages: "Show percentages",
    hidePercentages: "Hide percentages",
    pdfDownloadCaption: "PDF",
    xlsxDownloadCaption: "Excel",
    csvDownloadCaption: "CSV",
    saveDiagramAsPNG: "Download plot as a PNG file",
    hideEmptyAnswers: "Hide empty answers",
    showEmptyAnswers: "Show empty answers",
    "topNValueText-1": "All answers",
    "topNValueText5": "Top 5 answers",
    "topNValueText10": "Top 10 answers",
    "topNValueText20": "Top 20 answers",
    hideMissingAnswers: "Hide missing answers",
    showMissingAnswers: "Show missing answers",
    missingAnswersLabel: "Missing answers",
    noVisualizerForQuestion: "This question type is not visualized yet",
    noResults: "There are no results yet",
    showPerValues: "Per Values",
    showPerColumns: "Per Columns",
    answer: "Answer",
    correctAnswer: "Correct answer: ",
    percent: "Percent",
    percentage: "Percentage",
    statistics_chart: "Chart",
    responses: "Responses",
    visualizer_nps: "NPS",
    npsScore: "NPS",
    npsPromoters: "Promoters",
    npsPassives: "Passives",
    npsDetractors: "Detractors",
    axisXSelectorTitle: "X axis:",
    axisYSelectorTitle: "Y axis:"
};
// Uncomment the lines below if you create a custom dictionary.
// Replace "en" with a custom locale code (for example, "fr" or "de"),
// Replace `englishStrings` with the name of the variable that contains the custom dictionary.
// localization.locales["en"] = englishStrings;
// localization.localeNames["en"] = "English";

var localization = {
    currentLocaleValue: "",
    defaultLocaleValue: "en",
    locales: {},
    localeNames: {},
    supportedLocales: [],
    get currentLocale() {
        return this.currentLocaleValue === this.defaultLocaleValue
            ? ""
            : this.currentLocaleValue;
    },
    set currentLocale(val) {
        this.currentLocaleValue = val;
    },
    get defaultLocale() {
        return this.defaultLocaleValue;
    },
    set defaultLocale(val) {
        this.defaultLocaleValue = val;
    },
    getString: function (strName) {
        var loc = this.currentLocale
            ? this.locales[this.currentLocale]
            : this.locales[this.defaultLocale];
        if (!loc || !loc[strName])
            loc = this.locales[this.defaultLocale];
        var result = loc[strName];
        if (result === undefined) {
            result = this.locales["en"][strName] || strName;
        }
        return result;
    },
    getLocales: function () {
        var res = [];
        res.push("");
        if (this.supportedLocales && this.supportedLocales.length > 0) {
            for (var i = 0; i < this.supportedLocales.length; i++) {
                res.push(this.supportedLocales[i]);
            }
        }
        else {
            for (var key in this.locales) {
                res.push(key);
            }
        }
        res.sort();
        return res;
    }
};
var surveyStrings = englishStrings;
localization.locales["en"] = englishStrings;
localization.localeNames["en"] = "English";

class DocumentHelper {
    static createSelector(options, isSelected, handler, title) {
        const selectWrapper = document.createElement("div");
        selectWrapper.className = "sa-question__select-wrapper";
        if (title) {
            const titleElement = DocumentHelper.createElement("span", "sa-question__select-title", {
                innerText: title,
            });
            selectWrapper.appendChild(titleElement);
        }
        const select = document.createElement("select");
        select.className = "sa-question__select";
        options.forEach((option) => {
            let optionElement = DocumentHelper.createElement("option", "", {
                value: option.value,
                text: option.text,
                selected: isSelected(option),
            });
            select.appendChild(optionElement);
        });
        select.onchange = handler;
        selectWrapper.appendChild(select);
        return selectWrapper;
    }
    static createButton(handler, text = "", className = "sa-toolbar__button") {
        const button = DocumentHelper.createElement("span", className, {
            innerText: text,
            onclick: handler,
        });
        return button;
    }
    static createElement(tagName, className = "", attrs) {
        var el = document.createElement(tagName);
        el.className = className;
        if (!!attrs) {
            Object.keys(attrs).forEach(function (key) {
                el[key] = attrs[key];
            });
        }
        return el;
    }
    static createSvgElement(path) {
        const svgElem = document.createElementNS("http://www.w3.org/2000/svg", "svg");
        const useElem = document.createElementNS("http://www.w3.org/2000/svg", "use");
        useElem.setAttributeNS("http://www.w3.org/1999/xlink", "href", "#sa-svg-" + path);
        svgElem.appendChild(useElem);
        return svgElem;
    }
    static createSvgButton(path) {
        const btn = (DocumentHelper.createElement("button", "sa-table__svg-button"));
        btn.appendChild(DocumentHelper.createSvgElement(path));
        return btn;
    }
    static createSvgToggleButton(svgPath1, svPpath2, text1, text2, handler1, handler2, state = "first", className = "sa-toolbar__button sa-toolbar__svg-button") {
        const svg1 = DocumentHelper.createSvgElement(svgPath1);
        const svg2 = DocumentHelper.createSvgElement(svPpath2);
        const button = DocumentHelper.createElement("button", className);
        const toggle = (e) => {
            if (state === "first") {
                state = "second";
                button.title = text2;
                button.removeChild(svg1);
                button.appendChild(svg2);
                handler2(e);
            }
            else if (state === "second") {
                state = "first";
                button.title = text1;
                button.removeChild(svg2);
                button.appendChild(svg1);
                handler1(e);
            }
        };
        if (state === "first") {
            button.title = text1;
            button.appendChild(svg1);
        }
        else if ((state = "second")) {
            button.title = text2;
            button.appendChild(svg2);
        }
        button.onclick = toggle;
        return button;
    }
    static createInput(className, placeholder = "", defaultValue = "") {
        var el = DocumentHelper.createElement("input", className, {
            placeholder: placeholder,
            defaultValue: defaultValue,
        });
        return el;
    }
}
function createCommercialLicenseLink() {
    const container = DocumentHelper.createElement("div", "sa-commercial");
    const link = DocumentHelper.createElement("a", "sa-commercial__text", {
        href: "https://www.surveyjs.io/Buy",
        target: "_blank",
    });
    const containerSpan = DocumentHelper.createElement("span", "");
    const icon = DocumentHelper.createSvgElement("noncommercial");
    const textSpan = DocumentHelper.createElement("span", "sa-commercial__product", {
        innerText: "Please purchase a SurveyJS Analytics developer license to use it in your app.",
    });
    container.appendChild(link).appendChild(containerSpan);
    containerSpan.appendChild(icon);
    containerSpan.appendChild(textSpan);
    return container;
}
function createLoadingIndicator() {
    const container = DocumentHelper.createElement("div", "sa-data-loading-indicator-panel");
    const loadingIndicator = DocumentHelper.createElement("div", "sa-data-loading-indicator");
    loadingIndicator.innerHTML = `
      <svg width="64" height="64" viewBox="0 0 64 64" fill="none" xmlns="http://www.w3.org/2000/svg">
        <g clip-path="url(#clip0_17928_11482)">
          <path d="M32 64C14.36 64 0 49.65 0 32C0 14.35 14.36 0 32 0C49.64 0 64 14.35 64 32C64 49.65 49.64 64 32 64ZM32 4C16.56 4 4 16.56 4 32C4 47.44 16.56 60 32 60C47.44 60 60 47.44 60 32C60 16.56 47.44 4 32 4Z" fill="#E5E5E5"></path>
          <path d="M53.2101 55.2104C52.7001 55.2104 52.1901 55.0104 51.8001 54.6204C51.0201 53.8404 51.0201 52.5704 51.8001 51.7904C57.0901 46.5004 60.0001 39.4704 60.0001 31.9904C60.0001 24.5104 57.0901 17.4804 51.8001 12.1904C51.0201 11.4104 51.0201 10.1404 51.8001 9.36039C52.5801 8.58039 53.8501 8.58039 54.6301 9.36039C60.6701 15.4004 64.0001 23.4404 64.0001 31.9904C64.0001 40.5404 60.6701 48.5704 54.6301 54.6204C54.2401 55.0104 53.7301 55.2104 53.2201 55.2104H53.2101Z" fill="#19B394"></path>
        </g>
        <defs>
          <clipPath id="clip0_17928_11482">
            <rect width="64" height="64" fill="white"></rect>
          </clipPath>
        </defs>
      </svg>
      `;
    container.appendChild(loadingIndicator);
    return container;
}
class DataHelper {
    static zipArrays(...arrays) {
        let zipArray = [];
        for (let i = 0; i < arrays[0].length; i++) {
            zipArray[i] = [];
            arrays.forEach((arr) => {
                zipArray[i].push(arr[i]);
            });
        }
        return zipArray;
    }
    static unzipArrays(zipArray) {
        let arrays = [];
        zipArray.forEach((value, i) => {
            value.forEach((val, j) => {
                if (!arrays[j])
                    arrays[j] = [];
                arrays[j][i] = val;
            });
        });
        return arrays;
    }
    static sortDictionary(keys, values, desc) {
        let dictionary = this.zipArrays(keys, values);
        let comparator = (a, b, asc = true) => {
            let result = a[1] < b[1] ? 1 : a[1] == b[1] ? 0 : -1;
            return asc ? result : result * -1;
        };
        dictionary.sort((a, b) => {
            return desc ? comparator(a, b, false) : comparator(a, b);
        });
        let keysAndValues = this.unzipArrays(dictionary);
        return { keys: keysAndValues[0], values: keysAndValues[1] };
    }
    static toPercentage(value, maxValue) {
        return (value / maxValue) * 100;
    }
}
function createLinksContainer(links) {
    const linksContainer = DocumentHelper.createElement("div");
    links.forEach((link) => {
        linksContainer.appendChild(DocumentHelper.createElement("a", "", {
            innerText: link.name,
            download: link.name,
            href: link.content,
        }));
    });
    return linksContainer;
}
function createImagesContainer(links) {
    const linksContainer = DocumentHelper.createElement("div");
    links.forEach((link) => {
        linksContainer.appendChild(DocumentHelper.createElement("img", "", {
            alt: link.name,
            src: link.content,
        }));
    });
    return linksContainer;
}
function toPrecision(value, precision = 2) {
    const base = Math.pow(10, precision);
    return Math.round(base * value) / base;
}

// This dictionary contains 1 untranslated or inherited localization strings.
// These strings are commented out. Uncomment and edit them if you want to add your translations.
var farsiStrings = {
    groupButton: "گروه بندی با",
    ungroupButton: "حذف گروه",
    selectButton: "انتخاب",
    hideColumn: "مخفی کردن ستون",
    showColumn: "نمایش ستون",
    makePrivateColumn: "خصوصی کردن ستون",
    makePublicColumn: "عمومی کردن ستون",
    moveToDetail: "انتقال به جزئیات",
    showAsColumn: "نمایش به عنوان ستون",
    filterPlaceholder: "جستجو...",
    removeRows: "حذف سطرها",
    showLabel: "نمایش",
    entriesLabel: "ورودی",
    visualizer_text: "متون در جدول",
    visualizer_wordcloud: "ابر کلمات",
    visualizer_histogram: "هیستوگرام",
    visualizer_number: "متوسط",
    visualizer_choices: "جدول",
    visualizer_selectBase: "نمودار",
    // visualizer_matrix: "Chart",
    chartType_bar: "میله ای",
    chartType_vbar: "نوار عمودی",
    chartType_stackedbar: "نوار پشته",
    chartType_doughnut: "دونات",
    chartType_pie: "دابره ای(pie)",
    chartType_scatter: "پراکندگی(Scatter)",
    chartType_gauge: "عقربه ای",
    chartType_bullet: "Bullet",
    hideButton: "مخفی",
    makePrivateButton: "خصوصی کردن",
    makePublicButton: "عمومی کردن",
    showButton: "نمایش",
    filter: "فیلتر",
    resetFilter: "بازنشانی فیلترها",
    changeLocale: "تغییر محلی",
    clearButton: "پاک کردن",
    addElement: "انتخاب سوال برای نمایش...",
    defaultOrder: "ترتیب پیشفرض",
    ascOrder: "صعودی",
    descOrder: "نزولی",
    showMinorColumns: "نمایش ستونهای فرعی",
    otherCommentTitle: "سایر موارد و نظرات",
    showPercentages: "نمایش درصد",
    hidePercentages: "درصدهای مخفی",
    pdfDownloadCaption: "پی دی اف",
    xlsxDownloadCaption: "اکسل",
    csvDownloadCaption: "Csv",
    saveDiagramAsPNG: "دانلود طرح به عنوان یک فایل PNG",
    hideEmptyAnswers: "مخفی کردن پاسخهای خالی",
    showEmptyAnswers: "نمایش پاسخهای خالی",
    "topNValueText-1": "همه پاسخ ها",
    topNValueText5: "5 پاسخ برتر",
    topNValueText10: "10 پاسخ برتر",
    topNValueText20: "20 پاسخ برتر",
    hideMissingAnswers: "مخفی کردن پاسخهای گمشده",
    showMissingAnswers: "نمایش پاسخهای گمشده",
    missingAnswersLabel: "پاسخ های گمشده",
    noVisualizerForQuestion: "این نوع سوال هنوز تجسم نشده است",
    noResults: "هنوز نتیجه ای حاصل نشده است",
    showPerValues: "به ازای هر ارزش",
    showPerColumns: "در هر ستون",
    answer: "پاسخ",
    correctAnswer: "پاسخ صحیح: ",
    percent: "درصد",
    responses: "پاسخ"
};
localization.locales["fa"] = farsiStrings;
localization.localeNames["fa"] = "فارسی";
// The following strings have been translated by a machine translation service
// Remove those strings that you have corrected manually
// undefined.makePrivateColumn: "Make column private" => "خصوصی کردن ستون"
// undefined.makePublicColumn: "Make column public" => "عمومی کردن ستون"
// undefined.filterPlaceholder: "Search..." => "جستجو..."
// undefined.removeRows: "Remove rows" => "حذف سطرها"
// undefined.showLabel: "Show" => "نمایش"
// undefined.entriesLabel: "entries" => "ورودی"
// undefined.visualizer_histogram: "Histogram" => "هیستوگرام"
// undefined.visualizer_number: "Average" => "متوسط"
// undefined.visualizer_choices: "Table" => "جدول"
// undefined.visualizer_selectBase: "Chart" => "نمودار"
// undefined.chartType_vbar: "Vertical Bar" => "نوار عمودی"
// undefined.chartType_stackedbar: "Stacked Bar" => "نوار پشته"
// undefined.chartType_doughnut: "Doughnut" => "دونات"
// undefined.makePrivateButton: "Make private" => "خصوصی کردن"
// undefined.makePublicButton: "Make public" => "عمومی کردن"
// undefined.showButton: "Show" => "نمایش"
// undefined.filter: "Filter" => "فیلتر"
// undefined.changeLocale: "Change Locale" => "تغییر محلی"
// undefined.defaultOrder: "Default Order" => "ترتیب پیشفرض"
// undefined.ascOrder: "Ascending" => "صعودی"
// undefined.descOrder: "Descending" => "نزولی"
// undefined.showMinorColumns: "Show minor columns" => "نمایش ستونهای فرعی"
// undefined.otherCommentTitle: "Other items and comments" => "سایر موارد و نظرات"
// undefined.showPercentages: "Show percentages" => "نمایش درصد"
// undefined.hidePercentages: "Hide percentages" => "درصدهای مخفی"
// undefined.pdfDownloadCaption: "PDF" => "پی دی اف"
// undefined.xlsxDownloadCaption: "Excel" => "اکسل"
// undefined.csvDownloadCaption: "CSV" => "Csv"
// undefined.saveDiagramAsPNG: "Download plot as a PNG file" => "دانلود طرح به عنوان یک فایل PNG"
// undefined.hideEmptyAnswers: "Hide empty answers" => "مخفی کردن پاسخهای خالی"
// undefined.showEmptyAnswers: "Show empty answers" => "نمایش پاسخهای خالی"
// undefined.topNValueText-1: "All answers" => "همه پاسخ ها"
// undefined.topNValueText5: "Top 5 answers" => "5 پاسخ برتر"
// undefined.topNValueText10: "Top 10 answers" => "10 پاسخ برتر"
// undefined.topNValueText20: "Top 20 answers" => "20 پاسخ برتر"
// undefined.hideMissingAnswers: "Hide missing answers" => "مخفی کردن پاسخهای گمشده"
// undefined.showMissingAnswers: "Show missing answers" => "نمایش پاسخهای گمشده"
// undefined.missingAnswersLabel: "Missing answers" => "پاسخ های گمشده"
// undefined.noVisualizerForQuestion: "This question type is not visualized yet" => "این نوع سوال هنوز تجسم نشده است"
// undefined.noResults: "There are no results yet" => "هنوز نتیجه ای حاصل نشده است"
// undefined.showPerValues: "Per Values" => "به ازای هر ارزش"
// undefined.showPerColumns: "Per Columns" => "در هر ستون"
// undefined.answer: "Answer" => "پاسخ"
// undefined.correctAnswer: "Correct answer: " => "پاسخ صحیح: "
// undefined.percent: "Percent" => "درصد"
// undefined.responses: "Responses" => "پاسخ"

// This dictionary contains 1 untranslated or inherited localization strings.
// These strings are commented out. Uncomment and edit them if you want to add your translations.
var frenchStrings = {
    groupButton: "Grouper",
    ungroupButton: "Dissocier",
    selectButton: "Sélectionner",
    hideColumn: "Masquer colonne",
    showColumn: "Afficher colonne",
    makePrivateColumn: "Rendre la colonne privé",
    makePublicColumn: "Rendre la colonne public",
    moveToDetail: "Déplacer vers détails",
    showAsColumn: "Afficher en colonne",
    filterPlaceholder: "Rechercher...",
    removeRows: "Supprimer colonne",
    showLabel: "Afficher",
    entriesLabel: "entrées",
    visualizer_text: "Textes en table",
    visualizer_wordcloud: "Nuages de mots",
    visualizer_histogram: "Histogramme",
    visualizer_number: "Moyenne",
    visualizer_choices: "Table",
    visualizer_selectBase: "Graphique",
    // visualizer_matrix: "Chart",
    chartType_bar: "Barres",
    chartType_vbar: "Barre verticale",
    chartType_stackedbar: "Barres empilées",
    chartType_doughnut: "Anneau",
    chartType_pie: "Secteurs",
    chartType_scatter: "Nuages de points",
    chartType_gauge: "Gauge",
    chartType_bullet: "Bulles",
    hideButton: "Masquer",
    makePrivateButton: "Rendre privé",
    makePublicButton: "Rendre publique",
    showButton: "Afficher",
    filter: "Filtre",
    resetFilter: "Rafraichir Filtres",
    changeLocale: "Changer les paramètres régionaux",
    clearButton: "Rafraichir",
    addElement: "Choisir la question à afficher...",
    defaultOrder: "Defaut",
    ascOrder: "Ascendant",
    descOrder: "Descendant",
    showMinorColumns: "Afficher les colonnes mineures",
    otherCommentTitle: "Autre élément et commentaires",
    showPercentages: "Afficher les pourcentages",
    hidePercentages: "Masquer les pourcentages",
    pdfDownloadCaption: "PDF",
    xlsxDownloadCaption: "Exceller",
    csvDownloadCaption: "Fichier CSV",
    saveDiagramAsPNG: "Sauver en png",
    hideEmptyAnswers: "Masquer les réponses vides",
    showEmptyAnswers: "Afficher les réponses vides",
    "topNValueText-1": "Toutes les réponses",
    topNValueText5: "Top 5 des réponses",
    topNValueText10: "Top 10 des réponses",
    topNValueText20: "Top 20 des réponses",
    hideMissingAnswers: "Masquer les réponses manquantes",
    showMissingAnswers: "Afficher les réponses manquantes",
    missingAnswersLabel: "Réponses manquantes",
    noVisualizerForQuestion: "Ce type de question n'est pas encore visualisé",
    noResults: "Il n'y a pas encore de résultats",
    showPerValues: "Par valeur",
    showPerColumns: "Par colonne",
    answer: "Réponse",
    correctAnswer: "Réponse correcte: ",
    percent: "Pourcentage",
    responses: "Réponses"
};
localization.locales["fr"] = frenchStrings;
localization.localeNames["fr"] = "Français";
// The following strings have been translated by a machine translation service
// Remove those strings that you have corrected manually
// undefined.visualizer_histogram: "Histogram" => "Histogramme"
// undefined.visualizer_number: "Average" => "Moyenne"
// undefined.visualizer_choices: "Table" => "Table"
// undefined.visualizer_selectBase: "Chart" => "Graphique"
// undefined.chartType_vbar: "Vertical Bar" => "Barre verticale"
// undefined.changeLocale: "Change Locale" => "Changer les paramètres régionaux"
// undefined.showPercentages: "Show percentages" => "Afficher les pourcentages"
// undefined.hidePercentages: "Hide percentages" => "Masquer les pourcentages"
// undefined.pdfDownloadCaption: "PDF" => "PDF"
// undefined.xlsxDownloadCaption: "Excel" => "Exceller"
// undefined.csvDownloadCaption: "CSV" => "Fichier CSV"
// undefined.hideEmptyAnswers: "Hide empty answers" => "Masquer les réponses vides"
// undefined.showEmptyAnswers: "Show empty answers" => "Afficher les réponses vides"
// undefined.topNValueText-1: "All answers" => "Toutes les réponses"
// undefined.topNValueText5: "Top 5 answers" => "Top 5 des réponses"
// undefined.topNValueText10: "Top 10 answers" => "Top 10 des réponses"
// undefined.topNValueText20: "Top 20 answers" => "Top 20 des réponses"
// undefined.hideMissingAnswers: "Hide missing answers" => "Masquer les réponses manquantes"
// undefined.showMissingAnswers: "Show missing answers" => "Afficher les réponses manquantes"
// undefined.missingAnswersLabel: "Missing answers" => "Réponses manquantes"

// This dictionary contains 1 untranslated or inherited localization strings.
// These strings are commented out. Uncomment and edit them if you want to add your translations.
var germanStrings = {
    groupButton: "Gruppieren",
    ungroupButton: "Gruppierung aufheben",
    selectButton: "Mich auswählen",
    hideColumn: "Spalte ausblenden",
    showColumn: "Spalte einblenden",
    makePrivateColumn: "Spalte privat machen",
    makePublicColumn: "Spalte veröffentlichen",
    moveToDetail: "in Details verschieben",
    showAsColumn: "Als Spalte anzeigen",
    filterPlaceholder: "Suche...",
    removeRows: "Zeilen entfernen",
    showLabel: "anzeigen",
    entriesLabel: "Einträge",
    visualizer_text: "Texte in Tabellenform",
    visualizer_wordcloud: "Wordcloud",
    visualizer_histogram: "Histogramm",
    visualizer_number: "Durchschnitt",
    visualizer_choices: "Liste",
    visualizer_selectBase: "Diagrammtyp",
    // visualizer_matrix: "Chart",
    chartType_bar: "Balkendiagramm",
    chartType_vbar: "Balkendiagramm vertikal",
    chartType_stackedbar: "Gestapeltes Balkendiagramm",
    chartType_doughnut: "Donut-Diagramm",
    chartType_pie: "Kreisdiagramm",
    chartType_scatter: "Punktediagramm",
    chartType_gauge: "Pegeldiagramm",
    chartType_bullet: "Aufzählungsdiagramm",
    hideButton: "Ausblenden",
    makePrivateButton: "Privat machen",
    makePublicButton: "Öffentlich machen",
    showButton: "Anzeigen",
    filter: "Filter",
    resetFilter: "Filter zurücksetzen",
    changeLocale: "Lokalisierung ändern",
    clearButton: "Einstellungen löschen",
    addElement: "Wählen Sie eine Frage, um zu zeigen...",
    defaultOrder: "Standardreihenfolge",
    ascOrder: "Aufsteigend",
    descOrder: "Absteigend",
    showMinorColumns: "Kleinere Spalten anzeigen",
    otherCommentTitle: "Andere Punkte und Kommentare",
    showPercentages: "Prozentsätze anzeigen",
    hidePercentages: "Prozentsätze ausblenden",
    pdfDownloadCaption: "PDF",
    xlsxDownloadCaption: "Excel",
    csvDownloadCaption: "CSV",
    saveDiagramAsPNG: "Diagramm als png herunterladen",
    hideEmptyAnswers: "Leere Antworten ausblenden",
    showEmptyAnswers: "Leere Antworten anzeigen",
    "topNValueText-1": "Alle Antworten",
    topNValueText5: "Top 5 Antworten",
    topNValueText10: "Top 10 Antworten",
    topNValueText20: "Top 20 Antworten",
    hideMissingAnswers: "Fehlende Antworten ausblenden",
    showMissingAnswers: "Fehlende Antworten anzeigen",
    missingAnswersLabel: "Fehlede Antworten",
    noVisualizerForQuestion: "Für diesen Fragetyp gibt es noch keine Visualisierung.",
    noResults: "Es liegen noch keine Ergebnisse vor",
    showPerValues: "Pro Wert",
    showPerColumns: "Pro Spalte",
    answer: "Antwort",
    correctAnswer: "Korrekte Antwort: ",
    percent: "Prozent",
    responses: "Antworten"
};
localization.locales["de"] = germanStrings;
localization.localeNames["de"] = "Deutsch";

// This dictionary contains 1 untranslated or inherited localization strings.
// These strings are commented out. Uncomment and edit them if you want to add your translations.
var norwegianStrings = {
    groupButton: "Grupper",
    ungroupButton: "Opphev gruppering",
    selectButton: "Velg",
    hideColumn: "Skjul kolonne",
    showColumn: "Vis kolonne",
    makePrivateColumn: "Gjør kolonne privat",
    makePublicColumn: "Gjør kolonne offentlig",
    moveToDetail: "Flytt til detaljert visning",
    showAsColumn: "Vis som kolonne",
    filterPlaceholder: "Søk...",
    removeRows: "Fjern rader",
    showLabel: "Vis",
    entriesLabel: "oppføringer",
    visualizer_text: "Tabellvisning",
    visualizer_wordcloud: "Ordsky",
    visualizer_histogram: "Histogram",
    visualizer_number: "Gjennomsnitt",
    visualizer_choices: "Bord",
    visualizer_selectBase: "Sjøkart",
    // visualizer_matrix: "Chart",
    chartType_bar: "Stolpediagram",
    chartType_vbar: "Vertikal stolpe",
    chartType_stackedbar: "Stablet stolpediagram",
    chartType_doughnut: "Hjuldiagram",
    chartType_pie: "Sektordiagram",
    chartType_scatter: "Punktdiagram",
    chartType_gauge: "Målediagram",
    chartType_bullet: "Kulediagram",
    hideButton: "Skjul",
    makePrivateButton: "Gjør privat",
    makePublicButton: "Gjør offentlig",
    showButton: "Vis",
    filter: "Filter",
    resetFilter: "Nullstill filter",
    changeLocale: "Bytt språk",
    clearButton: "Tøm",
    addElement: "Velg spørsmål...",
    defaultOrder: "Standard",
    ascOrder: "Stigende",
    descOrder: "Synkende",
    showMinorColumns: "Vis mindre kolonner",
    otherCommentTitle: "Annet og kommentarer",
    showPercentages: "Vis prosenter",
    hidePercentages: "Skjul prosenter",
    pdfDownloadCaption: "PDF",
    xlsxDownloadCaption: "Overgå",
    csvDownloadCaption: "CSV",
    saveDiagramAsPNG: "Last ned plott som en PNG-fil",
    hideEmptyAnswers: "Skjul tomme svar",
    showEmptyAnswers: "Vis tomme svar",
    "topNValueText-1": "Alle svar",
    topNValueText5: "Topp 5 svar",
    topNValueText10: "Topp 10 svar",
    topNValueText20: "Topp 20 svar",
    hideMissingAnswers: "Skjul manglende svar",
    showMissingAnswers: "Vis manglende svar",
    missingAnswersLabel: "Mangler svar",
    noVisualizerForQuestion: "Denne spørsmålstypen er ikke visualisert ennå",
    noResults: "Det er ingen resultater ennå",
    showPerValues: "Per verdier",
    showPerColumns: "Per kolonner",
    answer: "Svare",
    correctAnswer: "Riktig svar: ",
    percent: "Prosent",
    responses: "Svar"
};
localization.locales["no"] = norwegianStrings;
localization.localeNames["no"] = "Norsk";
// The following strings have been translated by a machine translation service
// Remove those strings that you have corrected manually
// undefined.visualizer_histogram: "Histogram" => "Histogram"
// undefined.visualizer_number: "Average" => "Gjennomsnitt"
// undefined.visualizer_choices: "Table" => "Bord"
// undefined.visualizer_selectBase: "Chart" => "Sjøkart"
// undefined.chartType_vbar: "Vertical Bar" => "Vertikal stolpe"
// undefined.filter: "Filter" => "Filter"
// undefined.pdfDownloadCaption: "PDF" => "PDF"
// undefined.xlsxDownloadCaption: "Excel" => "Overgå"
// undefined.csvDownloadCaption: "CSV" => "CSV"
// undefined.saveDiagramAsPNG: "Download plot as a PNG file" => "Last ned plott som en PNG-fil"
// undefined.hideEmptyAnswers: "Hide empty answers" => "Skjul tomme svar"
// undefined.showEmptyAnswers: "Show empty answers" => "Vis tomme svar"
// undefined.topNValueText-1: "All answers" => "Alle svar"
// undefined.topNValueText5: "Top 5 answers" => "Topp 5 svar"
// undefined.topNValueText10: "Top 10 answers" => "Topp 10 svar"
// undefined.topNValueText20: "Top 20 answers" => "Topp 20 svar"
// undefined.hideMissingAnswers: "Hide missing answers" => "Skjul manglende svar"
// undefined.showMissingAnswers: "Show missing answers" => "Vis manglende svar"
// undefined.missingAnswersLabel: "Missing answers" => "Mangler svar"
// undefined.noVisualizerForQuestion: "This question type is not visualized yet" => "Denne spørsmålstypen er ikke visualisert ennå"
// undefined.noResults: "There are no results yet" => "Det er ingen resultater ennå"
// undefined.showPerValues: "Per Values" => "Per verdier"
// undefined.showPerColumns: "Per Columns" => "Per kolonner"
// undefined.answer: "Answer" => "Svare"
// undefined.correctAnswer: "Correct answer: " => "Riktig svar: "
// undefined.percent: "Percent" => "Prosent"
// undefined.responses: "Responses" => "Svar"

// This dictionary contains 1 untranslated or inherited localization strings.
// These strings are commented out. Uncomment and edit them if you want to add your translations.
var portugueseStrings = {
    groupButton: "Agrupar Por Mim",
    ungroupButton: "Desagrupar Por Mim",
    selectButton: "Selecionar",
    hideColumn: "Esconder coluna",
    showColumn: "Mostrar coluna",
    makePrivateColumn: "Tornar coluna privada",
    makePublicColumn: "Tornar coluna pública",
    moveToDetail: "Mover para Detalhes",
    showAsColumn: "Mostrar como Coluna",
    filterPlaceholder: "Pesquisar...",
    removeRows: "Remover linhas",
    showLabel: "Mostrar",
    entriesLabel: "entradas",
    visualizer_text: "Textos em tabela",
    visualizer_wordcloud: "Nuvem de palavras",
    visualizer_histogram: "Histograma",
    visualizer_number: "Média",
    visualizer_choices: "Tabela",
    visualizer_selectBase: "Gráfico",
    // visualizer_matrix: "Chart",
    chartType_bar: "Barra",
    chartType_vbar: "Barra vertical",
    chartType_stackedbar: "Barra Empilhada",
    chartType_doughnut: "Rosca",
    chartType_pie: "Tarte",
    chartType_scatter: "Lastro",
    chartType_gauge: "Escala",
    chartType_bullet: "Bala",
    hideButton: "Esconder",
    makePrivateButton: "Tornar privado",
    makePublicButton: "Tornar público",
    showButton: "Mostrar",
    filter: "Filtrar",
    resetFilter: "Restaurar Filtro",
    changeLocale: "Mudar Localização",
    clearButton: "Limpar",
    addElement: "Escolher questão para mostrar...",
    defaultOrder: "Ordenação Padrão",
    ascOrder: "Ascendente",
    descOrder: "Descendente",
    showMinorColumns: "Mostrar colunas menores",
    otherCommentTitle: "Outros itens e comentários",
    showPercentages: "Mostrar percentuais",
    hidePercentages: "Esconder percentuais",
    pdfDownloadCaption: "PDF",
    xlsxDownloadCaption: "Excel",
    csvDownloadCaption: "CSV",
    saveDiagramAsPNG: "Baixar diagrama como PNG",
    hideEmptyAnswers: "Esconder perguntas vazias",
    showEmptyAnswers: "Mostrar perguntas vazias",
    "topNValueText-1": "Todas as respostas",
    topNValueText5: "Primeiras 5 respostas",
    topNValueText10: "Primeiras 10 respostas",
    topNValueText20: "Primeiras 20 respostas",
    hideMissingAnswers: "Ocultar perguntas não respondidas",
    showMissingAnswers: "Mostrar perguntas não respondidas",
    missingAnswersLabel: "Perguntas não respondidas",
    noVisualizerForQuestion: "Este tipo de pergunta ainda não foi visualizada",
    noResults: "Ainda não possui resultados",
    showPerValues: "Por valores",
    showPerColumns: "Por colunas",
    answer: "Resposta",
    correctAnswer: "Resposta correta: ",
    percent: "Percentagem",
    responses: "Respostas"
};
//Uncomment these two lines on creating a translation file. You should replace "en" and enStrings with your locale ("fr", "de" and so on) and your variable.
localization.locales["pt"] = portugueseStrings;
localization.localeNames["pt"] = "Português";

var russianStrings = {
    groupButton: "Сгруппировать",
    ungroupButton: "Разгруппировать",
    selectButton: "Выбрать",
    hideColumn: "Скрыть столбец",
    showColumn: "Показать столбец",
    makePrivateColumn: "Сделать столбец приватным",
    makePublicColumn: "Сделать столбец публичным",
    moveToDetail: "Спрятать в раскрывающуюся секцию",
    showAsColumn: "Показать как столбец",
    filterPlaceholder: "Поиск...",
    removeRows: "Удалить строки",
    showLabel: "Показать",
    entriesLabel: "записей",
    visualizer_text: "Таблица",
    visualizer_wordcloud: "Облако тэгов",
    visualizer_histogram: "Гистограмма",
    visualizer_number: "Среднее значение",
    visualizer_choices: "Таблица",
    visualizer_selectBase: "Диаграмма",
    visualizer_matrix: "Диаграмма",
    chartType_bar: "Столбчатая диаграмма",
    chartType_vbar: "Вертикальная столбчатая диаграмма",
    chartType_stackedbar: "Столбчатая диаграмма с накоплением",
    chartType_doughnut: "Кольцевая диаграмма",
    chartType_pie: "Круговая диаграмма",
    chartType_scatter: "Точечная диаграмма",
    chartType_gauge: "Измерительная диаграмма",
    chartType_bullet: "Маркированная диаграмма",
    hideButton: "Скрыть",
    makePrivateButton: "Сделать приватным",
    makePublicButton: "Сделать публичным",
    showButton: "Показать",
    filter: "Фильтр",
    resetFilter: "Очистить фильтр",
    changeLocale: "Сменить язык",
    clearButton: "Очистить",
    addElement: "Выберите вопрос...",
    defaultOrder: "Порядок по умолчанию",
    ascOrder: "Сортировать по возрастанию",
    descOrder: "Сортировать по убыванию",
    showMinorColumns: "Показать второстепенные столбцы",
    otherCommentTitle: "Другое и комментарии",
    showPercentages: "Показать проценты",
    hidePercentages: "Скрыть проценты",
    pdfDownloadCaption: "PDF",
    xlsxDownloadCaption: "Excel",
    csvDownloadCaption: "CSV",
    saveDiagramAsPNG: "Скачать диаграмму в формате PNG",
    hideEmptyAnswers: "Скрыть пустые ответы",
    showEmptyAnswers: "Показать пустые ответы",
    "topNValueText-1": "Все ответы",
    topNValueText5: "Первые 5 ответов",
    topNValueText10: "Первые 10 ответов",
    topNValueText20: "Первые 20 ответов",
    hideMissingAnswers: "Скрыть отсутствующие ответы",
    showMissingAnswers: "Показать отсутствующие ответы",
    missingAnswersLabel: "Отсутствующие ответы",
    noVisualizerForQuestion: "Этот тип вопроса не поддерживается",
    noResults: "Результаты отсутствуют",
    showPerValues: "По значениям",
    showPerColumns: "По столбцам",
    answer: "Ответ",
    correctAnswer: "Правильный ответ: ",
    percent: "процентов",
    responses: "ответов"
};
localization.locales["ru"] = russianStrings;
localization.localeNames["ru"] = "Русский";

// This dictionary contains 1 untranslated or inherited localization strings.
// These strings are commented out. Uncomment and edit them if you want to add your translations.
var dutchStrings = {
    groupButton: "Groeperen op",
    ungroupButton: "Groepering verwijderen",
    selectButton: "Selecteer mij",
    hideColumn: "Kolom verbergen",
    showColumn: "Toon kolom",
    makePrivateColumn: "Kolom privé maken",
    makePublicColumn: "Maak kolom openbaar",
    moveToDetail: "Ga naar detail",
    showAsColumn: "Weergeven als kolom",
    filterPlaceholder: "Zoeken...",
    removeRows: "Verwijder rijen",
    showLabel: "Tonen",
    entriesLabel: "inzendingen",
    visualizer_text: "Teksten in tabel",
    visualizer_wordcloud: "Wordcloud",
    visualizer_histogram: "Histogram",
    visualizer_number: "Gemiddeld",
    visualizer_choices: "Tabel",
    visualizer_selectBase: "Grafiek",
    // visualizer_matrix: "Chart",
    chartType_bar: "Staafdiagram (liggend)",
    chartType_vbar: "Staafdiagram",
    chartType_stackedbar: "Staafdiagram (gestapeld)",
    chartType_doughnut: "Donut",
    chartType_pie: "Cirkeldiagram",
    chartType_scatter: "Spreidingsdiagram",
    chartType_gauge: "Meterdiagram",
    chartType_bullet: "Kogelgrafiek",
    hideButton: "Verbergen",
    makePrivateButton: "Maak prive",
    makePublicButton: "Openbaar maken",
    showButton: "Tonen",
    filter: "Filter",
    resetFilter: "Filter resetten",
    changeLocale: "Wijzig de landinstelling",
    clearButton: "Wissen",
    addElement: "Kies een vraag om weer te geven ...",
    defaultOrder: "Standaardvolgorde",
    ascOrder: "Oplopend",
    descOrder: "Aflopend",
    showMinorColumns: "Toon kleine kolommen",
    otherCommentTitle: "Andere items en opmerkingen",
    showPercentages: "Percentages weergeven",
    hidePercentages: "Verberg percentages",
    pdfDownloadCaption: "PDF",
    xlsxDownloadCaption: "Excel",
    csvDownloadCaption: "CSV",
    saveDiagramAsPNG: "Download plot als PNG",
    hideEmptyAnswers: "Verberg lege antwoorden",
    showEmptyAnswers: "Toon lege antwoorden",
    "topNValueText-1": "Alle antwoorden",
    topNValueText5: "Top 5 antwoorden",
    topNValueText10: "Top 10 antwoorden",
    topNValueText20: "Top 20 antwoorden",
    hideMissingAnswers: "Ontbrekende antwoorden verbergen",
    showMissingAnswers: "Ontbrekende antwoorden weergeven",
    missingAnswersLabel: "Ontbrekende antwoorden",
    noVisualizerForQuestion: "Dit vraagtype is nog niet gevisualiseerd",
    noResults: "Er zijn nog geen resultaten",
    showPerValues: "Per waarden",
    showPerColumns: "Per Kolommen",
    answer: "Antwoorden",
    correctAnswer: "Goed antwoord: ",
    percent: "Procent",
    responses: "Reacties"
};
//Uncomment these two lines on creating a translation file. You should replace "en" and enStrings with your locale ("fr", "de" and so on) and your variable.
localization.locales["nl"] = dutchStrings;
localization.localeNames["nl"] = "Nederlands";
// The following strings have been translated by a machine translation service
// Remove those strings that you have corrected manually
// undefined.visualizer_histogram: "Histogram" => "Histogram"
// undefined.visualizer_number: "Average" => "Gemiddeld"
// undefined.filter: "Filter" => "Filter"
// undefined.hideMissingAnswers: "Hide missing answers" => "Ontbrekende antwoorden verbergen"
// undefined.showMissingAnswers: "Show missing answers" => "Ontbrekende antwoorden weergeven"
// undefined.missingAnswersLabel: "Missing answers" => "Ontbrekende antwoorden"
// undefined.noVisualizerForQuestion: "This question type is not visualized yet" => "Dit vraagtype is nog niet gevisualiseerd"
// undefined.noResults: "There are no results yet" => "Er zijn nog geen resultaten"
// undefined.showPerValues: "Per Values" => "Per waarden"
// undefined.showPerColumns: "Per Columns" => "Per Kolommen"
// undefined.answer: "Answer" => "Antwoorden"
// undefined.correctAnswer: "Correct answer: " => "Goed antwoord: "
// undefined.percent: "Percent" => "Procent"
// undefined.responses: "Responses" => "Reacties"

// This dictionary contains 1 untranslated or inherited localization strings.
// These strings are commented out. Uncomment and edit them if you want to add your translations.
var spanishStrings = {
    groupButton: "Grupo Por Mí",
    ungroupButton: "Desagrupar por mí",
    selectButton: "Seleccionarme",
    hideColumn: "Ocultar columna",
    showColumn: "Mostrar columna",
    makePrivateColumn: "Hacer que la columna sea privada",
    makePublicColumn: "Hacer pública la columna",
    moveToDetail: "Mover al detalle",
    showAsColumn: "Mostrar como columna",
    filterPlaceholder: "Buscar ...",
    removeRows: "Eliminar filas",
    showLabel: "Mostrar",
    entriesLabel: "Entradas",
    visualizer_text: "Textos en tabla",
    visualizer_wordcloud: "Wordcloud",
    visualizer_histogram: "Histograma",
    visualizer_number: "Promedio",
    visualizer_choices: "Mesa",
    visualizer_selectBase: "Gráfico",
    // visualizer_matrix: "Chart",
    chartType_bar: "Bar",
    chartType_vbar: "Barra vertical",
    chartType_stackedbar: "Barra apilada",
    chartType_doughnut: "Donut",
    chartType_pie: "Pie",
    chartType_scatter: "Dispersión",
    chartType_gauge: "Gauge",
    chartType_bullet: "Bullet",
    hideButton: "Ocultar",
    makePrivateButton: "Hacer privado",
    makePublicButton: "Hacer público",
    showButton: "Mostrar",
    filter: "Filtro",
    resetFilter: "Restablecer filtro",
    changeLocale: "Cambiar configuración regional",
    clearButton: "Borrar",
    addElement: "Elija la pregunta para mostrar ...",
    defaultOrder: "Orden predeterminado",
    ascOrder: "Ascendente",
    descOrder: "Descendente",
    showMinorColumns: "Mostrar columnas secundarias",
    otherCommentTitle: "Otros elementos y comentarios",
    showPercentages: "Mostrar porcentajes",
    hidePercentages: "Ocultar porcentajes",
    pdfDownloadCaption: "PDF",
    xlsxDownloadCaption: "Excel",
    csvDownloadCaption: "CSV",
    saveDiagramAsPNG: "Descargar el gráfico como png",
    hideEmptyAnswers: "Ocultar respuestas vacías",
    showEmptyAnswers: "Mostrar respuestas vacías",
    "topNValueText-1": "Todas las respuestas",
    topNValueText5: "Las 5 respuestas principales",
    topNValueText10: "Las 10 respuestas principales",
    topNValueText20: "20 respuestas principales",
    hideMissingAnswers: "Ocultar las respuestas que faltan",
    showMissingAnswers: "Mostrar las respuestas que faltan",
    missingAnswersLabel: "Respuestas faltantes",
    noVisualizerForQuestion: "Este tipo de pregunta aún no se visualiza",
    noResults: "Aún no hay resultados",
    showPerValues: "Por valores",
    showPerColumns: "Por columnas",
    answer: "Respuesta",
    correctAnswer: "Respuesta correcta: ",
    percent: "Por ciento",
    responses: "Respuestas"
};
localization.locales["es"] = spanishStrings;
localization.localeNames["es"] = "Español";
// The following strings have been translated by a machine translation service
// Remove those strings that you have corrected manually
// undefined.groupButton: "Group By Me" => "Grupo Por Mí"
// undefined.entriesLabel: "entries" => "Entradas"
// undefined.visualizer_histogram: "Histogram" => "Histograma"
// undefined.visualizer_number: "Average" => "Promedio"
// undefined.visualizer_choices: "Table" => "Mesa"
// undefined.visualizer_selectBase: "Chart" => "Gráfico"
// undefined.chartType_vbar: "Vertical Bar" => "Barra vertical"
// undefined.hideMissingAnswers: "Hide missing answers" => "Ocultar las respuestas que faltan"
// undefined.showMissingAnswers: "Show missing answers" => "Mostrar las respuestas que faltan"
// undefined.missingAnswersLabel: "Missing answers" => "Respuestas faltantes"
// undefined.showPerValues: "Per Values" => "Por valores"
// undefined.showPerColumns: "Per Columns" => "Por columnas"
// undefined.answer: "Answer" => "Respuesta"
// undefined.correctAnswer: "Correct answer: " => "Respuesta correcta: "
// undefined.percent: "Percent" => "Por ciento"
// undefined.responses: "Responses" => "Respuestas"

// This dictionary contains 1 untranslated or inherited localization strings.
// These strings are commented out. Uncomment and edit them if you want to add your translations.
var italianStrings = {
    groupButton: "Raggruppa",
    ungroupButton: "Dividi",
    selectButton: "Selezionami",
    hideColumn: "Nascondi colonna",
    showColumn: "Mostra colonna",
    makePrivateColumn: "Rendi la colonna privata",
    makePublicColumn: "Rendi la colonna pubblica",
    moveToDetail: "Sposta in Dettaglio",
    showAsColumn: "Mostra come Colonna",
    filterPlaceholder: "Cerca...",
    removeRows: "Rimuovi righe",
    showLabel: "Mostra",
    entriesLabel: "voci",
    visualizer_text: "Parole in tabella",
    visualizer_wordcloud: "Nuvola di parole",
    visualizer_histogram: "Istogramma",
    visualizer_number: "Media",
    visualizer_choices: "Tabella",
    visualizer_selectBase: "Grafico",
    // visualizer_matrix: "Chart",
    chartType_bar: "Grafico a barre",
    chartType_vbar: "Grafico a barre verticali",
    chartType_stackedbar: "Grafico a barre sovrapposte",
    chartType_doughnut: "Grafico a ciambella",
    chartType_pie: "Grafico a torta",
    chartType_scatter: "Grafico a dispersione",
    chartType_gauge: "Grafico Gauge",
    chartType_bullet: "Punto",
    hideButton: "Nascondi",
    makePrivateButton: "Rendi privato",
    makePublicButton: "Rendi pubblico",
    showButton: "Mostra",
    filter: "Filtro",
    resetFilter: "Cancella filtro",
    changeLocale: "Cambia lingua",
    clearButton: "Cancella",
    addElement: "Scegli la domanda da mostrare...",
    defaultOrder: "Ordine di Default",
    ascOrder: "Ascendente",
    descOrder: "Discendente",
    showMinorColumns: "Mostra colonne secondarie",
    otherCommentTitle: "Altri punti e commenti",
    showPercentages: "Mostra percentuali",
    hidePercentages: "Nascondi percentuali",
    pdfDownloadCaption: "PDF",
    xlsxDownloadCaption: "Excel",
    csvDownloadCaption: "CSV",
    saveDiagramAsPNG: "Scarica il grafico in formato png",
    hideEmptyAnswers: "Nascondi risposte vuote",
    showEmptyAnswers: "Mostra risposte vuote",
    "topNValueText-1": "Tutte le risposte",
    topNValueText5: "Top 5 risposte",
    topNValueText10: "Top 10 risposte",
    topNValueText20: "Top 20 risposte",
    hideMissingAnswers: "Nascondi le risposte mancanti",
    showMissingAnswers: "Mostra le risposte mancanti",
    missingAnswersLabel: "Risposte mancanti",
    noVisualizerForQuestion: "Questo tipo di domanda non è ancora stato visualizzato",
    noResults: "Non ci sono ancora risultati",
    showPerValues: "Per Valori",
    showPerColumns: "Per Colonne",
    answer: "Risposta",
    correctAnswer: "Risposta esatta: ",
    percent: "Percentuale",
    responses: "Risposte"
};
localization.locales["it"] = italianStrings;
localization.localeNames["it"] = "Italiano";
// The following strings have been translated by a machine translation service
// Remove those strings that you have corrected manually
// undefined.correctAnswer: "Correct answer: " => "Risposta esatta: "

// This dictionary contains 1 untranslated or inherited localization strings.
// These strings are commented out. Uncomment and edit them if you want to add your translations.
var arabicStrings = {
    groupButton: "زرار المجموعة",
    ungroupButton: "تم التحديث من خلالى",
    selectButton: "اختر",
    hideColumn: "اخفاء العمود",
    showColumn: "اظهار",
    makePrivateColumn: "أجعل العمود خاص",
    makePublicColumn: "أجعل العمود عام",
    moveToDetail: "انتقل إلى التفاصيل",
    showAsColumn: "إظهار كعمود",
    filterPlaceholder: "بحث...",
    removeRows: "حذف",
    showLabel: "اظهار",
    entriesLabel: "مدخلات",
    visualizer_text: "جداول",
    visualizer_wordcloud: "كلمات",
    visualizer_histogram: "الرسم البياني",
    visualizer_number: "متوسط",
    visualizer_choices: "جدول",
    visualizer_selectBase: "رسم بياني",
    // visualizer_matrix: "Chart",
    chartType_bar: "شريط",
    chartType_vbar: "شريط عمودي",
    chartType_stackedbar: "شريط مكدس",
    chartType_doughnut: "مخطط حلقي",
    chartType_pie: "مخطط دائري",
    chartType_scatter: "مخطط التشتت",
    chartType_gauge: "مخطط القياس",
    chartType_bullet: "مخطط نقطي",
    hideButton: "اخفاء",
    makePrivateButton: "أجعل خاص",
    makePublicButton: "أجعل عام",
    showButton: "اظهار",
    filter: "ترشح",
    resetFilter: "إعادة تعيين",
    changeLocale: "تغيير اللغة",
    clearButton: "مسح",
    addElement: "اختر سؤالاً للعرض ...",
    defaultOrder: "الترتيب الافتراضي",
    ascOrder: "تصاعدى",
    descOrder: "تنازلى",
    showMinorColumns: "إظهار الأعمدة الثانوية",
    otherCommentTitle: "بنود وتعليقات أخرى",
    showPercentages: "عرض النسب المئوية",
    hidePercentages: "إخفاء النسب",
    pdfDownloadCaption: "PDF",
    xlsxDownloadCaption: "Excel",
    csvDownloadCaption: "CSV",
    saveDiagramAsPNG: "تنزيل بصيغة png",
    hideEmptyAnswers: "إخفاء الإجابات الفارغة",
    showEmptyAnswers: "إظهار الإجابات الفارغة",
    "topNValueText-1": "جميع الإجابات",
    topNValueText5: "أهم 5 إجابات",
    topNValueText10: "أفضل 10 إجابات",
    topNValueText20: "أفضل 20 إجابة",
    hideMissingAnswers: "إخفاء الإجابات المفقودة",
    showMissingAnswers: "عرض الإجابات المفقودة",
    missingAnswersLabel: "أجوبة مفقودة",
    noVisualizerForQuestion: "لم يتم العرض بعد",
    noResults: "لا يوجد نتائج للعرض",
    showPerValues: "لكل القيم",
    showPerColumns: "لكل الأعمدة",
    answer: "إجب",
    correctAnswer: "الاجابة الصحيحة: ",
    percent: "النسبه المئويه",
    responses: "الاستجابات"
};
localization.locales["ar"] = arabicStrings;
localization.localeNames["ar"] = "العربية";
// The following strings have been translated by a machine translation service
// Remove those strings that you have corrected manually
// undefined.topNValueText-1: "All answers" => "جميع الإجابات"
// undefined.topNValueText5: "Top 5 answers" => "أهم 5 إجابات"
// undefined.topNValueText10: "Top 10 answers" => "أفضل 10 إجابات"
// undefined.topNValueText20: "Top 20 answers" => "أفضل 20 إجابة"

var japaneseStrings = {
    groupButton: "グループ化",
    ungroupButton: "グループ解除",
    selectButton: "選択",
    hideColumn: "列を非表示",
    showColumn: "列を表示",
    makePrivateColumn: "列を非公開にする",
    makePublicColumn: "列を公開する",
    moveToDetail: "詳細に移動",
    showAsColumn: "列として表示",
    filterPlaceholder: "検索...",
    removeRows: "行を削除",
    showLabel: "表示",
    entriesLabel: "件",
    visualizer_text: "テキスト",
    visualizer_wordcloud: "ワードクラウド",
    visualizer_histogram: "ヒストグラム",
    visualizer_number: "平均",
    visualizer_choices: "テーブル",
    visualizer_selectBase: "チャート",
    chartType_bar: "棒グラフ",
    chartType_vbar: "縦棒グラフ",
    chartType_stackedbar: "積み上げ棒グラフ",
    chartType_doughnut: "ドーナツチャート",
    chartType_pie: "円グラフ",
    chartType_scatter: "散布図",
    chartType_gauge: "ゲージ",
    chartType_bullet: "バレット",
    hideButton: "非表示",
    makePrivateButton: "非公開にする",
    makePublicButton: "公開する",
    showButton: "表示",
    filter: "フィルター",
    resetFilter: "フィルターをリセット",
    changeLocale: "言語を変更",
    clearButton: "クリア",
    addElement: "表示する質問を選択...",
    defaultOrder: "デフォルト順",
    ascOrder: "昇順",
    descOrder: "降順",
    showMinorColumns: "マイナー列を表示",
    otherCommentTitle: "その他のアイテムとコメント",
    showPercentages: "パーセンテージを表示",
    hidePercentages: "パーセンテージを非表示",
    pdfDownloadCaption: "PDF",
    xlsxDownloadCaption: "Excel",
    csvDownloadCaption: "CSV",
    saveDiagramAsPNG: "プロットをPNGファイルとして保存",
    hideEmptyAnswers: "空の回答を非表示",
    showEmptyAnswers: "空の回答を表示",
    "topNValueText-1": "すべての回答",
    "topNValueText5": "上位5つの回答",
    "topNValueText10": "上位10の回答",
    "topNValueText20": "上位20の回答",
    hideMissingAnswers: "欠損値を非表示",
    showMissingAnswers: "欠損値を表示",
    missingAnswersLabel: "欠損値",
    noVisualizerForQuestion: "この質問タイプはまだ視覚化されていません",
    noResults: "まだ結果がありません",
    showPerValues: "値ごとに表示",
    showPerColumns: "列ごとに表示",
    answer: "回答",
    correctAnswer: "正解: ",
    percent: "パーセント",
    responses: "回答数"
};
localization.locales["ja"] = japaneseStrings;
localization.localeNames["ja"] = "Japanese";

// This dictionary contains 1 untranslated or inherited localization strings.
// These strings are commented out. Uncomment and edit them if you want to add your translations.
var plStrings = {
    groupButton: "Grupuj według mnie",
    ungroupButton: "Rozgrupuj według mnie",
    selectButton: "Wybierz",
    hideColumn: "Ukryj kolumnę",
    showColumn: "Pokaż kolumnę",
    makePrivateColumn: "Uczyń kolumnę prywatną",
    makePublicColumn: "Uczyń kolumnę publiczną",
    moveToDetail: "Przenieś do szczegółów",
    showAsColumn: "Pokaż jako kolumnę",
    filterPlaceholder: "Szukaj...",
    removeRows: "Usuń wiersze",
    showLabel: "Pokaż",
    entriesLabel: "wejścia",
    visualizer_text: "Teksty w tabeli",
    visualizer_wordcloud: "Chmura słów",
    visualizer_histogram: "Histogram",
    visualizer_number: "Średnia",
    visualizer_choices: "Tabela",
    visualizer_selectBase: "Wykres",
    visualizer_matrix: "Wykres",
    chartType_bar: "Słupkowy",
    chartType_vbar: "Słupkowy pionowy",
    chartType_stackedbar: "Słupkowy skumulowany",
    chartType_doughnut: "Oponka",
    chartType_pie: "Tarta",
    chartType_scatter: "Rozrzut",
    chartType_gauge: "Skala",
    chartType_bullet: "Pocisk",
    hideButton: "Ukryj",
    makePrivateButton: "Uczyń prywatnym",
    makePublicButton: "Uczyń publicznym",
    showButton: "Pokaż",
    filter: "Filtruj",
    resetFilter: "Resetuj filtr",
    changeLocale: "Zmień lokalizację",
    clearButton: "Wyczyść",
    addElement: "Wybierz pytanie do pokazania...",
    defaultOrder: "Domyślne sortowanie",
    ascOrder: "Rosnąco",
    descOrder: "Malejąco",
    showMinorColumns: "Pokaż mniejsze kolumny",
    otherCommentTitle: "Inne elementy i komentarze",
    showPercentages: "Pokaż procenty",
    hidePercentages: "Ukryj procenty",
    pdfDownloadCaption: "PDF",
    xlsxDownloadCaption: "Excel",
    csvDownloadCaption: "CSV",
    saveDiagramAsPNG: "Pobierz diagram jako PNG",
    hideEmptyAnswers: "Ukryj puste odpowiedzi",
    showEmptyAnswers: "Pokaż puste odpowiedzi",
    "topNValueText-1": "Wszystkie odpowiedzi",
    topNValueText5: "Pierwsze 5 odpowiedzi",
    topNValueText10: "Pierwsze 10 odpowiedzi",
    topNValueText20: "Pierwsze 20 odpowiedzi",
    hideMissingAnswers: "Ukryj nieodpowiedziane pytania",
    showMissingAnswers: "Pokaż nieodpowiedziane pytania",
    missingAnswersLabel: "Nieodpowiedziane pytania",
    noVisualizerForQuestion: "Ten typ pytania nie został jeszcze zwizualizowany",
    noResults: "Nie ma jeszcze wyników",
    showPerValues: "Według wartości",
    showPerColumns: "Według kolumn",
    answer: "Odpowiedź",
    correctAnswer: "Odpowiedź poprawna: ",
    percent: "Procent",
    responses: "Odpowiedzi"
};
//Uncomment these two lines on creating a translation file. You should replace "en" and enStrings with your locale ("fr", "de" and so on) and your variable.
localization.locales["pl"] = plStrings;
localization.localeNames["pl"] = "Polish";

// Uncomment the line below if you create a custom dictionary
var finnishStrings = {
    groupButton: "Erottele minulla",
    ungroupButton: "Poista erottele minulla",
    selectButton: "Valitse minut",
    hideColumn: "Piilota sarake",
    showColumn: "Näytä sarake",
    makePrivateColumn: "Merkitse sarake yksityiseksi",
    makePublicColumn: "Merkitse sarake julkiseksi",
    moveToDetail: "Siirrä yksityiskohtiin",
    showAsColumn: "Näytä sarakkeena",
    filterPlaceholder: "Etsi...",
    removeRows: "Poista rivejä",
    showLabel: "Näytä",
    entriesLabel: "vastauksia",
    visualizer_text: "Tekstitaulukko",
    visualizer_wordcloud: "Sanapilvi",
    visualizer_histogram: "Histogrammi",
    visualizer_number: "Keskiarvo",
    visualizer_choices: "Taulukko",
    visualizer_selectBase: "Kuvaaja",
    visualizer_matrix: "Kuvaaja",
    chartType_bar: "Pylväs",
    chartType_vbar: "Vaakapyöväs",
    chartType_stackedbar: "Pinottu pylväs",
    chartType_doughnut: "Donitsi",
    chartType_pie: "Piirakka",
    chartType_scatter: "Hajonta",
    chartType_gauge: "Mittari",
    chartType_bullet: "Luoti",
    chartType_line: "Viiva",
    hideButton: "Piilota",
    makePrivateButton: "Merkise yksityiseksi",
    makePublicButton: "Merkitse julkiseksi",
    showButton: "Näytä",
    filter: "Rajaa",
    resetFilter: "Nollaa Rajaus",
    changeLocale: "Vaihda kieli",
    clearButton: "Nollaa",
    addElement: "Valitse näytettävä kysymys...",
    defaultOrder: "Oletusjärjestys",
    ascOrder: "Nouseva",
    descOrder: "Laskeva",
    showMinorColumns: "Näytä alasarakkeet",
    otherCommentTitle: "Muut asiat ja kommentit",
    showPercentages: "Näytä prosentit",
    hidePercentages: "Piilota prosentit",
    pdfDownloadCaption: "PDF",
    xlsxDownloadCaption: "Excel",
    csvDownloadCaption: "CSV",
    saveDiagramAsPNG: "Lataa kuvaaja PNG kuvana",
    hideEmptyAnswers: "Piilota tyhjät vastaukset",
    showEmptyAnswers: "Näytä tyhjät vastaukset",
    "topNValueText-1": "Kaikki vastaukset",
    "topNValueText5": "Top 5 vastaukset",
    "topNValueText10": "Top 10 vastaukset",
    "topNValueText20": "Top 20 vastaukset",
    hideMissingAnswers: "Piilota puuttuvat vastaukset",
    showMissingAnswers: "Näytä puuttuvat vastaukset",
    missingAnswersLabel: "Puuttuvat vastaukset",
    noVisualizerForQuestion: "Tätä kysymystyyppiä ei voida visualisoida.",
    noResults: "Ei vielä tuloksia",
    showPerValues: "Näytä arvottain",
    showPerColumns: "Näytä sarakkeittain",
    answer: "Vastaus",
    correctAnswer: "Oikea vastaus: ",
    percent: "Prosentti",
    percentage: "Prosenttiyksikkö",
    statistics_chart: "Kuvaaja",
    responses: "Vastaukset",
    visualizer_nps: "NPS",
    npsScore: "NPS",
    npsPromoters: "Suosittelija",
    npsPassives: "Passiivinen",
    npsDetractors: "Arvostelija",
};
// Uncomment the lines below if you create a custom dictionary.
// Replace "en" with a custom locale code (for example, "fr" or "de"),
// Replace `englishStrings` with the name of the variable that contains the custom dictionary.
localization.locales["fi"] = finnishStrings;
localization.localeNames["fi"] = "Suomi";

var iconsData = {
	"detail": "<svg version=\"1.1\" xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 16 16\"><g><circle cx=\"1.5\" cy=\"8.5\" r=\"1.5\"></circle><circle cx=\"7.5\" cy=\"8.5\" r=\"1.5\"></circle><circle cx=\"13.5\" cy=\"8.5\" r=\"1.5\"></circle></g></svg>",
	"drag": "<svg version=\"1.1\" xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 16 16\"><g><polygon points=\"13,5 12,6 13,7 9,7 9,3 10,4 11,3 8,0 5,3 6,4 7,3 7,7 3,7 4,6 3,5 0,8 3,11 4,10 3,9 7,9 7,13 6,12 5,13 8,16 11,13 10,12 9,13 9,9 13,9 12,10 13,11 16,8 \"></polygon></g></svg>",
	"hide": "<svg version=\"1.1\" xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 16 16\"><g><path d=\"M12.79,4.2L16,1l-1-1l-3.32,3.32C10.57,2.55,9.32,2,8,2C3.63,2,0,7.97,0,7.97s1.27,2.1,3.21,3.82l-3.23,3.23 L0.96,16l3.35-3.34C5.41,13.44,6.67,14,8,14c4.37,0,8-6,8-6S14.73,5.91,12.79,4.2z M2,7.97c1.07-1.47,3.61-4,6-4 c0.8,0,1.6,0.27,2.35,0.68L9.57,5.43C9.11,5.14,8.58,4.97,8,4.97c-1.66,0-3,1.34-3,3c0,0.58,0.17,1.11,0.46,1.57l-0.97,0.97 C3.36,9.64,2.45,8.59,2,7.97z M8,11.97c-0.8,0-1.59-0.27-2.33-0.67l0.78-0.77C6.9,10.8,7.43,10.97,8,10.97c1.66,0,3-1.34,3-3 c0-0.57-0.17-1.09-0.44-1.55l0.98-0.98c1.11,0.87,2.01,1.91,2.46,2.52C12.93,9.44,10.39,11.97,8,11.97z\"></path></g></svg>",
	"makeprivate": "<svg version=\"1.1\" id=\"Layer_1\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" x=\"0px\" y=\"0px\" viewBox=\"0 0 16 16\" style=\"enable-background:new 0 0 16 16;\" xml:space=\"preserve\"><g><circle cx=\"8\" cy=\"4\" r=\"3\"></circle><path d=\"M8,8c-3.31,0-6,2.73-6,4.67S5.27,15,8,15s6-0.4,6-2.33S11.31,8,8,8z\"></path></g></svg>",
	"makepublic": "<svg version=\"1.1\" id=\"Layer_1\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" x=\"0px\" y=\"0px\" viewBox=\"0 0 16 16\" style=\"enable-background:new 0 0 16 16;\" xml:space=\"preserve\"><g><circle cx=\"6\" cy=\"5\" r=\"3\"></circle><path d=\"M10,5c0,0.34-0.06,0.67-0.14,0.99C9.91,5.99,9.95,6,10,6c1.66,0,3-1.34,3-3s-1.34-3-3-3 C8.97,0,8.07,0.52,7.53,1.3C8.98,1.91,10,3.33,10,5z\"></path><path d=\"M6,9c-3.31,0-6,2.73-6,4.67S3.27,16,6,16s6-0.4,6-2.33S9.31,9,6,9z\"></path><path d=\"M10,7C9.81,7,9.62,7.01,9.43,7.03C9.09,7.6,8.62,8.07,8.06,8.42c3,0.83,5.23,3.31,5.42,5.28 c1.44-0.31,2.52-0.92,2.52-2.03C16,9.74,13.31,7,10,7z\"></path></g></svg>",
	"movetodetails": "<svg version=\"1.1\" xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 16 16\"><g><circle cx=\"1.5\" cy=\"14.5\" r=\"1.5\"></circle><circle cx=\"7.5\" cy=\"14.5\" r=\"1.5\"></circle><circle cx=\"13.5\" cy=\"14.5\" r=\"1.5\"></circle><rect x=\"12\" y=\"1\" width=\"3\" height=\"9\"></rect><path d=\"M2.02,7.02L1,6L0,7l3,3l3-3L5,6L4.06,6.94C4.33,4.73,6.21,3,8.5,3C8.67,3,8.83,3.03,9,3.05V1.03 C8.83,1.01,8.67,1,8.5,1C5.08,1,2.27,3.66,2.02,7.02z\"></path></g></svg>",
	"noncommercial": "<svg version=\"1.1\" id=\"Layer_1\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\" x=\"0px\" y=\"0px\" viewBox=\"0 0 16 16\" style=\"enable-background:new 0 0 16 16;\" xml:space=\"preserve\"><path d=\"M8,0L0,15h16L8,0z M9,13H7v-2h2V13z M7,10V5h2v5H7z\"></path></svg>",
	"sorting": "<svg version=\"1.1\" xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 16 16\"><g><polygon points=\"8,3 10,5 11,4 8,1 5,4 6,5 \"></polygon><polygon points=\"8,13 6,11 5,12 8,15 11,12 10,11\"></polygon></g></svg>"
};

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
//@ts-ignore
const iconPrefix = "sa-svg-";
function getIconSymbolTemplate(iconId, iconSvg) {
    const startStr = "<svg ";
    const endStr = "</svg>";
    iconSvg = iconSvg.trim();
    const str = iconSvg.toLowerCase();
    if (str.substring(0, startStr.length) === startStr &&
        str.substring(str.length - endStr.length, str.length) === endStr) {
        return "<symbol " +
            "id=\"" + iconPrefix + iconId + "\" " +
            iconSvg.substring(startStr.length, str.length - endStr.length) +
            "</symbol>";
    }
}
const iconsHtml = Object.keys(iconsData).map(iconId => getIconSymbolTemplate(iconId, iconsData[iconId]));
const svgTemplate = `<svg style="display:none;">${iconsHtml}<svg>`;

export { DocumentHelper as D, DataHelper as a, createLinksContainer as b, createImagesContainer as c, createCommercialLicenseLink as d, svgTemplate as e, createLoadingIndicator as f, localization as l, surveyStrings as s, toPrecision as t };
